### PR TITLE
Infrastructure "compose.py refactor"

### PR DESCRIPTION
**Pull request by int-ua**
#### Summary

SUMMARY: None

#### Purpose of change

Fixing failure to compose when tileset directory contains at least one grayscale image (ImageMagick automatically optimizes images and it's a good thing) that turned into a major refactor.

#### Describe the solution

:warning: Please squash the commits on merging, suggested commit description follows: 

```
compose.py refactor

* convert images to `srgb` if they are not `srgb`
* changes to variable and method names
* replacement of `.format` with f-strings
* replacement of double quotes with single quotes
* type hints
* get rid of splitting images list into rows
* parametrize number of sprites across
* store tileset reference in Tilesheet objects
* move processing of file types into own methods
* no more empty starting images on all tilesheets, only global index 0 has to be ignored
* `--obsolete-fillers`: opt-in warning about obsoleted fillers
* warning about empty tile entries
* `tileset.txt` is required now in either source or output directory
* `tile_config.json` name no longer hardcoded, instead read from `tileset.txt`
* filling the last row is not required, it's done automatically by the `arrayjoin`
* wapcaplet: Embellish script docstring with usage examples 
```

#### Describe alternatives you've considered

Not replacing double quotes. But I prefer single quotes because they are less visually loading and require one less action to type.

Leaving `.format` in some places.

#### Testing

```sh
git clone --depth 1 https://github.com/I-am-Erk/CDDA-Tilesets.git
cd Cataclysm-DDA
mkdir ../UltimateCataclysm
mkdir ../UltimateCataclysm_new

git checkout master
~/.venv/bin/python3 tools/gfx_tools/compose.py --use-all ../CDDA-Tilesets/gfx/UltimateCataclysm ../UltimateCataclysm

git checkout compose
~/.venv/bin/python3 tools/gfx_tools/compose.py --use-all --obsolete-fillers ../CDDA-Tilesets/gfx/UltimateCataclysm ../UltimateCataclysm_new
```
But they **will** differ because I've removed the unnecessary addition of empty starting images to every sheet.

Also tested by composing all tilesets in https://github.com/I-am-Erk/CDDA-Tilesets and the https://github.com/pixel-32/CDDA-tileset/

#### Additional context

https://github.com/I-am-Erk/CDDA-Tilesets/runs/1613802647

### CODE REVIEW

File: tools/gfx_tools/compose.py [hunk 1]
```diff
@@ -371,109 +379,107 @@ def create_sheet(self, merge_pngs):
...
+        if not sheet.is_standard(tileset):
+            sheet_conf['sprite_width'] = sheet.sprite_width
+            sheet_conf['sprite_height'] = sheet.sprite_height
+            sheet_conf['sprite_offset_x'] = sheet.offset_x
+            sheet_conf['sprite_offset_y'] = sheet.offset_y
 
-        tiles_new_dict[ts_data.max_index] = ts_conf
+        tiles_new_dict[sheet.max_index] = sheet_conf
 
     # fing unused images
```
Comment (actual-nh) [inline tools/gfx_tools/compose.py:453]:
```suggestion
    # find unused images
```

**Comment by wapcaplet** [comment]
While I don't know tilesets well enough to fully evaluate this, the code is certainly more readable and idiomatic than before. I ran it on a few of the official CDDA tilesets from I-am-Erk's repo, and the results look OK.

I submitted a PR to your fork with a small improvement to the usage notes. Otherwise I can't think of any changes to suggest. Thanks for working on this!

File: tools/gfx_tools/compose.py [hunk 2]
```diff
@@ -198,163 +164,295 @@ def find_unused(self, use_all=False):
...
+    def write_composite_png(self) -> bool:
+        '''
+        Compose and save tilesheet PNG if there are sprites to work with
+        '''
+        if not self.sprites:
+            return False
+
+        # count empty spaces in the last row
+        self.tileset.pngnum += self.sheet_width - \
+            (len(self.sprites) % self.sheet_width)
```
Comment (int-ua) [inline tools/gfx_tools/compose.py:319]: Wrong, see #46618

**Comment by Fris0uman** [comment]
*`--obsolete-fillers`: opt-in warning about obsoleted fillers
* `tileset.txt` is required now in either source or output directory
* `tile_config.json` name no longer hardcoded, instead read from `tileset.txt`
Those and the `--use-all` option should probably be documented in `Tilesets.md`

**Pull request merged**